core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

```diff
@@ -0,0 +1,251 @@
+            podSpecs.updated(id, podSpec)
+          case None =>
+            podSpecs - id
+        }
+        val responseFromUpdate = SchedulerLogic.handlePodSpecUpdate(podSpecs, podRecords)(id, newState, ResponseBuilder.empty)
+
+        HandleSpecMessageResponse(
+          newPodSpecs,
+          Set(id),
+          responseFromUpdate)
+
+      case ReservationSpecChange(id, _) =>
+        throw new NotImplementedError("ReservationSpec support not yet implemented")
+    }
+  }
+
+  case class HandleMesosEventResponse(response: ResponseBuilder, changedPodIds: Set[PodId])
+  object HandleMesosEventResponse {
+    val empty = HandleMesosEventResponse(ResponseBuilder.empty, Set.empty)
+  }
+
+  def handleMesosEvent(podStatuses: Map[PodId, PodStatus], podSpecs: Map[PodId, PodSpec], cachedPendingLaunch: Set[PodId])
+                      (event: Mesos.Event): HandleMesosEventResponse = {
+    event match {
+      case offer: Mesos.Event.Offer =>
+        val changedPodIds = cachedPendingLaunch
+        val response = matchOffer(podSpecs)(offer, cachedPendingLaunch, ResponseBuilder.empty)
+        HandleMesosEventResponse(response, changedPodIds)
+
+      case Mesos.Event.StatusUpdate(taskId, taskStatus) =>
+
+        val podId = podIdFor(taskId)
+
+        if (podSpecs.contains(podId)) {
+          val newState = podStatuses.get(podId) match {
+            case Some(status) =>
+              status.copy(taskStatuses = status.taskStatuses.updated(taskId, taskStatus))
+            case None =>
+              PodStatus(podId, Map(taskId -> taskStatus))
+          }
+
+          val response = ResponseBuilder.empty.withChangeMessage(PodStatusChange(podId, Some(newState)))
+          HandleMesosEventResponse(
+            response,
+            Set(podId))
+        } else {
+          HandleMesosEventResponse.empty
+        }
+    }
+  }
+
+  /**
+    * We remove a task if it is not reachable and running, and it has no podSpec defined
+    *
+    * Should be called with the effects already applied for the specified podIds
+    * @param podIds podIds changed during the last frame
+    * @return
+    */
+  def pruneTaskStatuses(podStatuses: Map[PodId, PodStatus], podSpecs: Map[PodId, PodSpec])
+                       (podIds: Set[PodId]): ResponseBuilder = {
+    var b = ResponseBuilder.empty
+    def terminalOrUnreachable(status: PodStatus): Boolean = {
+      // Cheesy method
+      status.taskStatuses.values.forall(status => status == Mesos.TaskStatus.TASK_RUNNING)
+    }
+
+    podIds.foreach { podId =>
+      if (podStatuses.contains(podId)) {
+        val podSpecDefined = !podSpecs.contains(podId)
+        // prune terminal statuses for which there's no defined podSpec
+        if (!podSpecDefined && terminalOrUnreachable(podStatuses(podId))) {
+          b = b.withChangeMessage(PodStatusChange(podId, None))
+        }
+      }
+    }
+    b
+  }
+
+  private def podIdFor(taskId: TaskId): PodId = PodId(taskId)
+}
+
+class SchedulerLogic {
```

Comment:
I am still struggling - when do you put a method here and when in `SchedulerLogic` object? In class are only functions that modify the state?